### DISCUSSION TITLE
How to define script1 node attribute for child of child of child of child node (and so on)

### DISCUSSION THREAD

**Discussion by metacoma** [freeplane/freeplane#791] [Help] [answered]
Hello,

TLDR; i do some research with freeplane, and I have faced with groovy trouble

My question is how to properly define attribute script1 for node of child of child of child of child, here is my code:

```groovy
// Here i use ' (apostrof) to define multiline script
node[script1] = '''
   def second = node.createChild('second')
   //Here i need to use another sort of quotes, double quotes " 
   second['script1'] = """ 
       def third = node.createChild("third")
       // Here I need to use \' apostrof with backslahes
       third['script1'] = \'\'\'
           def fourth = node.createChild("fourth")
           // What sort of quotes i need use here?
           fourth['script1'] = ???
               def fifth = node.createChild("fifth")
               fifth['script1'] = ????
                  def sixth = ...
                      // and so on .....
               ???
           ???
       \'\'\'
       
   ""'
"''
```

This "ugly" code i generated by code generator (jinja2 templating), so my question how to properly define scripts for nodes inside child of child of child ...? 

Or, maybe, can anyone suggest another approach? There is only one condition - it must be one groovy file.

Thank you!

**Comment by macmarrum** [comment, marked as answer]
What you are looking for is a function, e.g.
```groovy
def createChildWithThisFunctionInScript1(_node) {
  def newNode = _node.createChild()
  newNode['script1'] = 'createChildWithThisFunctionInScript1(node)'
}
```

The function needs to be made available in every node. To do this in Freeplane, you can create a library -> https://docs.freeplane.org/attic/old-mediawiki-content/Your_own_utility_script_library.html

In essence, create a file in `{user-dir}/lib/`, e.g. named `FP.groovy`.
_Note: You can use `Tools->Open user directory`_
Write your function(s) there, with the `static` keyword, i.e.

```groovy
def static createChildWithThisFunctionInScript1(_node) {
  def newNode = _node.createChild()
  newNode['script1'] = 'FP.createChildWithThisFunctionInScript1(node)'
```

Restart Freeplane.

In your initial node add a script1 with the content `FP.createChildWithThisFunctionInScript1(node)`.

___

See also
- https://github.com/freeplane/freeplane/discussions/763#discussioncomment-3926185

**Comment by metacoma** [comment]
Hey @macmarrum 

First of all, thank you for your detailed explanation and groovy examples, it's saved a lot of my time :+1: 

Here a short demo of my freeplane research, i think you might be interested


https://user-images.githubusercontent.com/5146707/200052330-9e6de6e9-4648-4a3a-ab5f-b73adabc5a05.mp4
